### PR TITLE
Annotate and adjust more of CoreLib for trimming

### PR DESCRIPTION
**Pull request by MichalStrehovsky**

### CODE REVIEW

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/CustomAttribute.cs [hunk 1]
```diff
@@ -591,6 +591,8 @@ internal static bool IsDefined(ICustomAttributeProvider obj, Type attributeType,
         [MethodImplAttribute(MethodImplOptions.InternalCall)]
         private static extern bool IsDefinedInternal(ICustomAttributeProvider obj, Type AttributeType);
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2006:UnrecognizedReflectionPattern",
+            Justification = "Custom attributes are analyzed by the linker")]
```
Comment (vitek-karas) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/CustomAttribute.cs:595]: The justification is weird - maybe a bit more detail (linker analyzes all properties and marks them?)

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs [hunk 2]
```diff
@@ -1819,6 +1821,8 @@ public GenericTypeParameterBuilder[] DefineGenericParameters(params string[] nam
             return generic_params;
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2006:UnrecognizedReflectionPattern",
+            Justification = "Type.GetConstructor(ConstructorInfo) is not actually unsafe")]
```
Comment (vitek-karas) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs:1825]: Should we fix linker to not mark it as problematic then?

**Review by vitek-karas — APPROVED** [review]

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs (hunk 2 above)
Comment (MichalStrehovsky) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs:1825]: This overload of GetConstructor is a Monoism that doesn't exist elsewhere.

The question is whether linker should check whether the signatures match the expectations based on the public surface area.

I don't know if I really want us to spend time checking all the signatures...

Comment (vitek-karas) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs:1825]: OK - I'm fine with that - just add a comment why linker doesn't handle this "correctly"

File: src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs [hunk 3]
```diff
@@ -115,10 +116,11 @@ public virtual Type GetDefaultReader()
         // Returns the preferred IResourceWriter class for this kind of ResourceSet.
         // Subclasses of ResourceSet using their own Readers &; should override
         // GetDefaultReader and GetDefaultWriter.
+        // TODO: https://github.com/mono/linker/issues/943
+        [DynamicDependency(DynamicallyAccessedMemberTypes.PublicConstructors, "System.Resources.ResourceWriter", "System.Resources.Writer")]
         public virtual Type GetDefaultWriter()
         {
-            Assembly resourceWriterAssembly = Assembly.Load("System.Resources.Writer, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a");
-            return resourceWriterAssembly.GetType("System.Resources.ResourceWriter", throwOnError: true)!;
+            return Type.GetType("System.Resources.ResourceWriter, System.Resources.Writer", throwOnError: true)!;
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs:123]: @jkotas split this into the 2 calls in [https://github.com/dotnet/coreclr/pull/11314/files#diff-37ee6a4a589ffae33ce341f9c1a28133R136](https://github.com/dotnet/coreclr/pull/11314/files#diff-37ee6a4a589ffae33ce341f9c1a28133R136). Are we sure this is safe to collapse it back into a single `Type.GetType` call?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs:123]: The PR was reconciling differences between multiple forked copies of the same code. I do not think there was a strong reason why I have changed this to `Assembly.Load` + `GetType` that was used in the other fork.

It should be ok to change it back to `Type.GetType`.

**Review by stephentoub — APPROVED** [review]
Other than the existing comments, LGTM

**Review by eerhardt — APPROVED** [review]
:shipit:

**Comment by jkotas** [comment]
Failures are https://github.com/dotnet/runtime/issues/39484

**Pull request merged**